File: src/main/java/data/stack/RemoveKDigits_402.java
package data.stack;

import java.util.Stack;
import java.util.stream.Collectors;

public class RemoveKDigits_402 {
  public static String removeKDigits(String num, int k) {
    int len = num.length(), n = len - k;
    var stack = new Stack<Character>();
    if (k >= len) {
      return "0";
    }

    for (var i = 0; i < len; i++) {
      while (n < stack.size() + len - i && !stack.isEmpty() && stack.peek() > num.charAt(i)) {
        stack.pop();
      }
      if (stack.size() < n) {
        stack.push(num.charAt(i));
      }
    }

    var result = stack.stream().map(String::valueOf).collect(Collectors.joining());
    var i = 0;
    for (; i < result.length(); i++) {
      if (result.charAt(i) != '0') {
        break;
      }
    }
    return i == result.length() ? "0" : result.substring(i);
  }

  public static void main(String[] args) {
    System.out.println(removeKDigits("1432219", 3));
    System.out.println(removeKDigits("10200", 1));
    System.out.println(removeKDigits("100200", 1));
    System.out.println(removeKDigits("10", 2));
    System.out.println(removeKDigits("9", 1));
    System.out.println(removeKDigits("112", 1));
    System.out.println(removeKDigits("10001", 4));
  }
}

File: src/main/kotlin/data/structure/stack/RemoveKDigits_402.kt
package data.structure.stack

import java.util.Stack

class KtRemoveKDigits_402 {
    // the same as FindTheMostCompetitiveSubsequence_1673
    fun removeKdigits(num: String, k: Int): String {
        val maxStack = Stack<Char>()
        val len = num.length - k  // keep len digits in final result
        if (len == 0) {
            return "0"
        }
        num.indices.forEach { i ->
            val c = num[i]
            while (maxStack.isNotEmpty() && maxStack.peek() > c && maxStack.size + num.length - i > len) {
                maxStack.pop()
            }
            if (maxStack.size < len) {
                maxStack.push(c)
            }
        }

        val digits = maxStack.joinToString("")
        var start = 0
        while (start < digits.length && digits[start] == '0') {
            start++
        }
        // return "0" if all digits are 0
        return if (start == digits.length) "0" else digits.substring(start)
    }
}

fun main() {
    val remove = KtRemoveKDigits_402()
    println(remove.removeKdigits("1432219", 3))
    println(remove.removeKdigits("10200", 1))
    println(remove.removeKdigits("10", 2))
}